### ISSUE TITLE
bcachefs detects but does not repair file corruption.

### ISSUE THREAD

**Issue by b-r-o-w-n** [koverstreet/bcachefs#146]
kernel: 5.7.0-g86fa1258a

tools: v0.1-227-g21ade39

I'm experimenting with a "metadata_replicas=2,data_replicas=2,metadata_replicas_required=2,data_replicas_required=2" setup.

I'm manually corrupting one of the copies of a file (/dev/sda1). The copy on /dev/sdb1 is correct.

The filesystem mounts ok and the file can be dumped and its contents are correct.

The fs detects the corrupted file on /dev/sda1 (it always seems to read /dev/sda1 first) and then uses the good copy on /dev/sdb1.

It generates this dmesg:

[ 621.287148] bcachefs (df0f537e-dcca-443a-82ea-e46bc4ca6f02): IO error on sda1 for data checksum error, inode 4098 offset 0: expected 0:6dfad270 got 0:5227d903 (type 5)

But, it never fixes the corruption.  I also cannot seem to force it to fix the corruption manually (or I do not know how to)
If I unmount the filesystem and remount I get the same error on file read (ie.. corruption was not fixed)

One nice thing would be some error counters in /sys/fs/bcachefs/...../ to be pegged. If there are error counters please let me know their names.

thanks

**Comment by ManoaNosea** [comment]
I got a similar problem, one of the drives was disconnected by accident and I had to restart the server, it show a lot of errors when tryed to unmount but after the restart when I run fsck on it, detected a journal error and asked to fix, when fix was attempted the error was "cannot allocate memory (-22)" and then it sayd "try with older version of bcachefs-tools" - but I never used any other version of bcachefs tools before, this is the only one...

**Comment by ManoaNosea** [comment]
mount of the filesystem worked without a problem and so far the data on it appears to be ok :)

**Comment by ManoaNosea** [comment]
```
[434589.637042] bcachefs (6318e365-e711-4f56-824d-96e4eccf8dec): IO error on sdi for data read; I/O
[434591.076847] bcachefs (6318e365-e711-4f56-824d-96e4eccf8dec): IO error on sdi for btree read: I/O
[434591.076850] bcachefs (6318e365-e711-4f56-824d-96e4eccf8dec): retrying read
```
and it goes on and on and on

**Comment by ManoaNosea** [comment]
it couldn't continue working because it doesn't detect that the same drive now have a new name (not sdi)

**Comment by Lyamc** [comment]
Hey, just wanted to let you know that this is in the TODO:

> TODO:
>  - scrub, repair of replicated data when one of the replicas fail the checksum
>    check
>  **- erasure coding needs repair code (it'll do reconstruct reads, but we don't have code to rewrite bad blocks in a stripe yet. this is going to be a hassle until we get backpointers)**
>  - fsck isn't checking refcounts of reflinked extents yet
>  - bcachefs tests in ktest need to be moved to xfstests
>  - user docs are still very minimal

**Comment by YellowOnion** [comment]
@Lyamc This has nothing to do with Erasure codes

I had similar issues.

```
$ bcachefs show-super /dev/disk/by-id/<blah> 
```

This showed me that one of my devices state was "read only".

changing it to readwrite with:

```
$ echo "readwrite" > /sys/fs/bcachefs/<UUID>/dev-<num>/state
```

then run:

```
bcachefs fsck -p <disk-0> <disk-1> ...
```

This should fix the "silent" fails trying to repair the drive.

Bug report should be called "fsck should work on readonly drives".

**Comment by b-r-o-w-n** [comment]
Has any progress been made on this issue?

I'm not sure I agree with "YellowOnion" above but I'm not qualified to comment as I do not know how its supposed to work.

In the above test  I have a 2 copy fs. When I corrupt the first copy the second copy is used.  No error counters are incremented.
It just moves to the next available copy or generates an i/o error because the error is not recoverable. 
 
So, the question is: is the design going to fix it on  the fly when its detected or wail till some kind of "scrub" operation/command 
is triggered sometime later? 

When you hit the checksum error in the code and find the good data you know the good data location and
the bad data location.

So, does anyone have an ball park estimate when an update might arrive? (No is a valid answer.)

**Comment by koverstreet** [comment]
No ETA yet - focusing on getting snapshots done and merged. Once that's
done I'll be able to look at other things, and this is something that
really shouldn't take much work.

**Comment by b-r-o-w-n** [comment]
Why did it get closed?

**Comment by YellowOnion** [comment]
It's for a very outdated version, I'm just closing anything that looks like a bug, and is over a year old since the code is in such a flux, I'll reopen it, since it looks like we need to implement scrub (still).

**Comment by b-r-o-w-n** [comment]
Hi,

Any planned progress here? 
I'm not sure that this is an enhancement.
Its like btrfs without the scrub command. You detect bit rot or corruption
and skip to the second or third copy but offer no tools to fix the problem.

**Comment by koverstreet** [comment]
Planned, but I've got higher priority items on my plate right now

**Comment by pavinjosdev** [comment]
@koverstreet Please add scrub support 🙏
I have a 4-drive RAID 1 setup using USB drives and errors are frequent if there's some disconnection or timeout waiting for writes to slower drives.

**Comment by koverstreet** [comment]
This is done in 6.11

**Comment by pavinjosdev** [comment]
> This is done in 6.11

Thanks for the scrub support, would rebalance be also available in 6.11? 🥺
